codebasin/report.py

```diff
@@ -244,3 +249,72 @@ def clustering(output_name, setmap):
         fig.savefig(fp)
 
     return "\n".join(lines)
+
+
+def find_duplicates(codebase: CodeBase) -> list[set[Path]]:
+    """
+    Search for duplicate files in the code base.
+
+    Returns
+    -------
+    list[set[Path]]
+        A list of all sets of Paths with identical contents.
+    """
+    # Search for possible matches using a hash, ignoring symlinks.
+    possible_matches = defaultdict(set)
```

Comment:
I'm very conflicted about how I feel about this pattern: L271 just acts like the `set` exists already but it creates the key/value upon access. On the one hand the behavior is nice, but in terms of readability...

I'm okay to merge this if you are?

My alternative suggestion actually is to be more verbose about it:

```python
if digest not in possible_matches:
    possible_matches[digest] = set()
possible_matches[digest].add(path)
```

it's more verbose but IMHO the expectation is clearer since the lines are next to each other

Reply:
I agree with you.  I've been burned by `defaultdict` enough times at this point that I don't think we should use it unless we have to.  The consumer of the `defaultdict` in this case doesn't rely on non-existent keys returning a default value, so I think it makes sense to make the change you've proposed here.  Fixed in 6a952ceb3c95be254ea757818935da3615b63390.